### ISSUE TITLE
qwen2训练loss出错

### ISSUE THREAD

**Issue by lambda-lee** [QwenLM/Qwen2#670]
使用qwen2、fp16 进行sft全参数微调的时候，有些token会输出nan, 导致loss为0。但推理的时候，使用fp16并不会有问题。这是什么原因呢，怎么解决，谢谢，之前使用qwen1.5 没有遇到过这个问题。

**Comment by jklj077** [comment]
Hi, which model were you finetuning? 

If Inf or NaN is countered, you will probably need to finetune it in bf16 or fp32.

**Comment by lambda-lee** [comment]
Qwen2-7B-Instruct

**Comment by Nidryen-zh** [comment]
I have same problem, using qwen2 fp16 SFT, the loss inside is nan, then print result is 0

**Comment by Nidryen-zh** [comment]
bf16 and fp32 is correct, however, V100 do not support bf16

**Comment by xsank** [comment]
@lambda-lee qwen2 你是用代码中的finetune脚本跑的吗

**Comment by lambda-lee** [comment]
> @lambda-lee qwen2 你是用代码中的finetune脚本跑的吗

基本是，有稍微的变动

**Comment by jklj077** [comment]
Hi, guys, 

I think the issue is quite clear, and it is very common with float16 training for LLMs. 

As previously replied, bfloat16 is the preferred data type for this model; if your device does not support the data type, you can try float32.

There are other options worth trying:
1. using a fused kernel implementation, e.g., pytorch sdpa (the memory efficient and the flash attention variants) or flash_attention_v2
2. other methods in dealing with training instability with float16 could help, e.g., using a lower learning rate
